android-sdk/src/main/java/org/sagebionetworks/bridge/android/util/ScheduledActivityUtil.java

```diff
@@ -0,0 +1,288 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ *
+ */
+
+package org.sagebionetworks.bridge.android.util;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.common.base.Function;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Multimaps;
+
+import org.sagebionetworks.bridge.rest.model.ScheduledActivity;
+import org.sagebionetworks.bridge.rest.model.TaskReference;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+
+/**
+ * Created by liujoshua on 2/7/2018.
+ */
+
+@SuppressWarnings("Convert2Lambda")
+public class ScheduledActivityUtil {
+
+    /**
+     * Function that returns the schedule plan guid for a scheduled activity.
+     */
+    public static final Function<ScheduledActivity, String> TO_SCHEDULE_PLAN_GUID =
+            new Function<ScheduledActivity, String>() {
+                @NonNull
+                @Override
+                public String apply(@NonNull ScheduledActivity scheduledActivity) {
+                    checkNotNull(scheduledActivity);
+
+                    return scheduledActivity.getSchedulePlanGuid();
+                }
+            };
+
+    /**
+     * Function that returns the task identifier for a scheduled activity, or null if the
+     * activity is a survey.
+     */
+    public static final Function<ScheduledActivity, String> TO_TASK_IDENTIFIER =
+            new Function<ScheduledActivity, String>() {
+                @Nullable
+                @Override
+                public String apply(@NonNull ScheduledActivity scheduledActivity) {
+                    checkNotNull(scheduledActivity);
+                    Verify.verify(scheduledActivity.getActivity() != null);
+
+                    TaskReference task = scheduledActivity.getActivity().getTask();
+                    if (task == null) {
+                        return null;
+                    }
+
+                    String taskId = task.getIdentifier();
+                    Verify.verify(taskId != null);
+                    return taskId;
+                }
+            };
+
+    /**
+     * Groups scheduled activities by schedule plan guid, maintaining order of scheduled activities.
+     * @param activities scheduled activities
+     * @return scheduled activities grouped by schedule plan guid
+     */
+    @NonNull
+    public static ImmutableMultimap<String, ScheduledActivity> groupBySchedulePlan(
+            @NonNull List<ScheduledActivity> activities) {
+        checkNotNull(activities);
+
+        return ImmutableMultimap.copyOf(
+                Multimaps.index(activities.iterator(), TO_SCHEDULE_PLAN_GUID)
+        );
+    }
+
+    // TODO: tests for iOS equivalents
+//    @VisibleForTesting
+//    static final Predicate<ScheduledActivity> HAS_FINISHED_ON =
```

Comment:
I never liked predicates on iOS because of their String-based format.  It was like you had to learn another language just to use and understand them.  This java version of predicate seems better.

Reply:
I had a hard time understanding the iOS predicates too